### PR TITLE
fix(app-router): keep isPending true across router.back() and router.forward()

### PR DESCRIPTION
**Pull request by NathanDrake2406**
## What this changes

`router.back()` and `router.forward()` now keep `useTransition().isPending` true from the synchronous call site until the RSC traversal commits. Matches the behavior shipped for `router.push` / `replace` in #868 and for RSC-level redirects in #870.

## Why

`router.back()` and `router.forward()` call `window.history.back/forward()` synchronously. The browser queues `popstate` as a new task, so by the time it fires the user's `React.startTransition` callback has already exited. No setState is tied to the transition, `useTransition().isPending` never latches, and any loading UI flashes to its idle state before the new route renders.

Wrapping the popstate handler itself in `startTransition` does not help, because that creates a transition unrelated to the one the user opened in their component.

## Approach

Arm a deferred `PendingBrowserRouterState` (a Promise published into the router's `useState` slot) inside the user's transition, then adopt it on the subsequent popstate's `navigateRsc`.

Three small production changes:

1. **`shims/navigation.ts`** - add `runProgrammaticTraversal(direction)`. It opens the user's `startTransition`, calls `window.__VINEXT_ARM_TRAVERSAL_PENDING__?.()`, then `history.back/forward()`. Gate the arm on `window.navigation.canGoBack` / `canGoForward` so a no-op traversal does not leave an unsettled pending.
2. **`server/app-browser-entry.ts`** - register `__VINEXT_ARM_TRAVERSAL_PENDING__` at hydration time; it invokes `beginPendingBrowserRouterState()`, which auto-settles any stale pending, publishes a fresh promise via the `setBrowserRouterState` setter, and records it as `activePendingBrowserRouterState`.
3. **`navigateRsc` adoption branch** - when `programmaticTransition` is false and an active pending exists, adopt it so the commit resolves the promise the user's transition is suspended on. User-initiated popstates (keyboard, gesture, address bar) find no active pending and fall through unchanged.

No registry, no id map, no safety timer. The Navigation API's synchronous availability check handles the no-op case; on older browsers without the Navigation API, arming is skipped and `isPending` preservation degrades to current Next.js behavior (no hang).

Non-goals: Pages Router is unaffected; the arm hook is defined only for App Router pages.

## Validation

- New E2E: `tests/e2e/app-router/nextjs-compat/router-back-forward-pending.spec.ts`, four specs using the MutationObserver idle-flash pattern from #870:
  - `router.back()` keeps `isPending` true until page A commits (reproduces the current bug pre-fix with log `["b:idle","b:pending","b:idle","__removed__"]`)
  - `router.forward()` keeps `isPending` true until page B commits
  - rapid `router.back(); router.back();` stays pending through the second hop
  - `router.back()` then `router.forward()` across separate clicks round-trips cleanly
- All 66 nextjs-compat E2E tests pass, including the existing push/redirect pending specs from #868 and #870.
- `vp check` clean (format, lint, type-aware).

## Risks / follow-ups

- Browsers without the Navigation API: `isPending` preservation is skipped for traversals. Same behavior as Next.js today on those browsers, no hang.
- A user-initiated popstate interleaved with a programmatic traversal in flight will adopt the active pending. End state is always a real committed tree; only the "wrong URL under the programmatic caller's pending" edge case differs from perfect classification. Stamping `history.state` with a programmatic marker would close this; deliberately out of scope.

### CODE REVIEW

**Comment by NathanDrake2406** [comment]
Addressed the review. Three changes in 72331374:

1. **Replaced the shared counter with a unique-id registry.** Copilot's race analysis is correct: two rapid back() calls set the counter to 2, the first popstate decrements to 1 (which equals the first call's captured generation), and the first safety-net timer then falsely treats its traversal as "no popstate fired" and clears the counter. The second popstate classifies itself as non-programmatic, navigateRsc skips adopting the active pending state, and isPending flashes when the superseded first call's finally settles early.

   `__VINEXT_PROGRAMMATIC_TRAVERSAL__: number | undefined` is now `__VINEXT_PROGRAMMATIC_TRAVERSALS__: Map<number, () => void> | undefined`. Each back/forward allocates a monotonic id and registers (id, settle) in the map. The safety-net timer deletes its own id and, if the entry was still present, invokes settle. The popstate handler consumes the oldest registered id (FIFO matches browser popstate delivery order) and discards the settle function — navigateRsc settles the active pending state on commit. Unique ids cannot collide, so an earlier safety-net timer can never steal a later traversal's bookkeeping.

2. **Factored `back()` and `forward()` into a single `runProgrammaticTraversal` helper** so the two paths share one code path and cannot drift.

3. **Test comment fix** — updated the header to reference `#back-pending-state` / `#forward-pending-state` rather than the nonexistent `#pending-state`.

**Comment by NathanDrake2406** [comment]
Review findings addressed in two commits on this branch.

## B1 — Rapid back(); back(); flashing isPending false

Root cause: `beginPendingBrowserRouterState` (app-browser-entry.ts:169) auto-settled the first pending when the second traversal began, and the popstate handler adopted `activePendingBrowserRouterState` (a shared slot) rather than the pending specific to each traversal.

Fix:

- `beginPendingBrowserRouterState` no longer auto-settles a prior unsettled pending. Each caller is responsible for its own settlement via the `finally` block in `navigateRsc`.
- `__VINEXT_BEGIN_PENDING_TRAVERSAL__` now returns `{ settle, id }` (previously just `settle`). The entry stores each traversal's `PendingBrowserRouterState` in a new module-local `traversalPendingStates: Map<number, PendingBrowserRouterState>` keyed by `id`.
- The popstate handler now retrieves the specific `PendingBrowserRouterState` by traversal id and passes it to `navigateRsc` via a new optional `pendingToAdopt?: PendingBrowserRouterState` parameter (7th arg, only used from the popstate handler's direct call). Push/replace/refresh continue to use the `programmaticTransition` path unchanged.

Trace for `back(); back();`:
1. Call 1: `beginPendingBrowserRouterState()` creates pending1, `setter(pending1.promise)`, stores at `traversalPendingStates[id=0]`
2. Call 2: `beginPendingBrowserRouterState()` creates pending2, `setter(pending2.promise)`, stores at `traversalPendingStates[id=1]`. **pending1 is not settled.**
3. popstate for back1: consumes id=0, `pendingToAdopt=pending1`, `navigateRsc(..., pendingToAdopt=pending1)` commits with correct state, resolves pending1.promise.
4. No popstate for back2 (no-op): safety-net timer fires, calls `settle2()`, resolves pending2.promise with current state.
No idle flash occurs.

Files: `packages/vinext/src/server/app-browser-entry.ts`, `packages/vinext/src/global.d.ts`

## B2 — Tests for rapid concurrent traversals

Two new E2E tests added to `tests/e2e/app-router/nextjs-compat/router-back-forward-pending.spec.ts`:

- "rapid back(); back(); keeps useTransition pending without idle flash": uses a MutationObserver to record every text value of `#back-pending-state` and asserts no "idle" appears between first "pending" and element removal.
- "back(); forward(); to different URLs commits to the forward target": fires `back(); forward();` in one `startTransition` and asserts the final URL is the forward target (page B), not the back target.

Fixture buttons added to `back-client.tsx`: `#router-double-back-btn` and `#router-back-then-forward-btn`.

## N1 — Overlapping popstates drop reducer output

With B1 fixed, each traversal holds its own `PendingBrowserRouterState`. `resolvePendingBrowserRouterState` only short-circuits if `pending.settled`, and each traversal has a distinct object, so concurrent traversals cannot interfere with each other's settlement. The `back(); forward();` test exercises this: the back navigation is superseded (navId staleness) and its pending is settled in the finally block; the forward navigation commits successfully and its pending resolves with correct state.

## N2 — Module-scoped id under Vite multi-instance

Previously `navigation.ts` had `let nextProgrammaticTraversalId = 1`. Id generation is now owned entirely by `app-browser-entry.ts` (via `nextTraversalPendingId`). The browser entry is loaded once and is not subject to Vite multi-environment duplication. `runProgrammaticTraversal` uses the id returned by `__VINEXT_BEGIN_PENDING_TRAVERSAL__` directly, no local counter.

File: `packages/vinext/src/shims/navigation.ts`

## N3 — User-initiated popstate races programmatic traversal

Documented as a known limitation in the popstate handler comment (app-browser-entry.ts ~line 1308). The race requires a user-initiated back/forward to arrive in the brief window between `invokeHistoryApi()` and the popstate handler consuming the registry entry. The outcome is a brief isPending flash on an edge case. A clean fix would require synchronously stamping history state before the history API call; this is deferred.

## N4 — Silent no-op pre-hydration

`runProgrammaticTraversal` now emits `console.warn` in DEV (`import.meta.env.DEV` gate) when `__VINEXT_BEGIN_PENDING_TRAVERSAL__` is undefined (pre-hydration or Pages Router), pointing at the likely cause. The `?.` optional-chain fallback is preserved.

File: `packages/vinext/src/shims/navigation.ts`

**Comment by NathanDrake2406** [comment]
Force-pushed a full rewrite of this branch.

The previous revision had drifted against `upstream/main`: it reverted merged PRs #875 (`rsc-fetch-errors.spec.ts` and non-ok RSC fetch recovery) and #878 (`_GLOBAL_HYDRATION_CONTEXT_KEY` in `shims/navigation.ts`), and removed ~84 unrelated lines from `tests/shims.test.ts`. The registry / Map / id / safety-timer / dual-write approach it settled on was also more complex than necessary and had been patched twice in review to chase edge cases.

New branch is based on `upstream/main` at 54516468 and has two commits:

1. `test(app-router): add failing E2E for router.back/forward isPending continuity` - fixtures plus four Playwright specs. On `upstream/main` the first spec fails with `["b:idle","b:pending","b:idle","__removed__"]`.
2. `fix(app-router): keep isPending true across router.back() and router.forward()` - the three production-code changes described in the PR body.

Net change: ~64 lines of production code across `shims/navigation.ts`, `server/app-browser-entry.ts`, and `global.d.ts`.

### How each earlier review item dissolves

- **Dual-write in `dispatchBrowserTree.applyAction`**: removed. Single-pending adoption means there is no orphan case to heal.
- **`back(); back();` with a no-op second back**: structurally impossible here. The second `router.back()` calls `beginPendingBrowserRouterState`, which auto-settles the first pending with the current committed state and registers a fresh one as the sole `activePendingBrowserRouterState`. React's state slot latches onto the latest `setter(promise)` call.
- **50ms / 200ms safety timer**: gone. The Navigation API's `canGoBack` / `canGoForward` is a synchronous availability hint; when the traversal would be a no-op, we skip arming and no pending is left unsettled. No timer needed.
- **Per-call registry / FIFO ordering nits**: N/A. There is no registry.
- **Pre-hydration `console.debug` gating**: N/A. The arm call short-circuits with `?.()` when the hook is undefined; silent by design.

### Validation

- All 66 nextjs-compat E2E tests pass locally, including the 2 from #868 / #870 which cover the `programmaticTransition=true` push/replace/redirect paths.
- `vp check` clean.
- Pre-fix: first spec fails as documented above. Post-fix: all four specs pass.

I'll wait on CI before re-requesting review.

**Comment by NathanDrake2406** [comment]
I'm curious how much each review cost cloudflare. They're using API pricing yeah? 💀

File: packages/vinext/src/server/app-browser-entry.ts [hunk 1]
```diff
@@ -1087,6 +1101,17 @@ function bootstrapHydration(rscStream: ReadableStream<Uint8Array>): void {
...
+        activePendingBrowserRouterState !== null &&
+        !activePendingBrowserRouterState.settled
+      ) {
+        // Popstate path: the shim already armed a pending inside the user's
+        // `React.startTransition` (router.back / router.forward). Adopt it so
+        // the commit resolves the promise React's state slot is suspended on,
+        // keeping `useTransition().isPending` true across the async popstate
+        // task boundary. User-initiated popstates (keyboard, gesture, address
+        // bar) find no active pending and fall through unchanged.
+        pendingRouterState = activePendingBrowserRouterState;
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/server/app-browser-entry.ts]: Race traced and addressed. The shared-pending hazard is real: in `back(); back();` overlap, both `navigateRsc` invocations adopt the same `activePendingBrowserRouterState`, and the older one's stale-id `finally` runs `settlePendingBrowserRouterState` before the surviving one reaches `applyAction` — the surviving `applyAction` then observed `pending.settled` and no-op'd, leaving React's slot bound to the stale promise resolution.

Applied the dual-write you suggested in `e18139d5`. `applyAction` now always calls `setter(routerReducer(...))` after `resolvePendingBrowserRouterState`. In the common case the two writes batch into one transition commit; in the race case the setter rebinds React's slot to the correct concrete state and re-commits.

Comment in `applyAction` names the invariant so this isn't mistaken for dead code.

File: packages/vinext/src/server/app-browser-entry.ts [hunk 2]
```diff
@@ -581,18 +581,19 @@ function BrowserRoot({
...
+    window.__VINEXT_APP_ROUTER_READY__ = true;
     return () => {
       if (setBrowserRouterState === setTreeStateValue) {
         setBrowserRouterState = null;
       }
       if (browserRouterStateRef === stateRef) {
         browserRouterStateRef = null;
+        window.__VINEXT_APP_ROUTER_READY__ = false;
+        window.__VINEXT_ARM_TRAVERSAL_PENDING__ = undefined;
       }
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/server/app-browser-entry.ts:650]: Good catch on the StrictMode dev cycle. Fixed in `40b543a0` by moving the arm hook assignment into BrowserRoot's mount `useLayoutEffect` (alongside `setBrowserRouterState` / `browserRouterStateRef`) so mount and cleanup own its lifecycle as a single unit. The duplicate assignment in `bootstrapHydration` is dropped, and a comment there points at the new owner.

File: packages/vinext/src/shims/navigation.ts [hunk 3]
```diff
@@ -1239,6 +1239,58 @@ export async function navigateClientSide(
...
+function runProgrammaticTraversal(direction: "back" | "forward"): void {
+  React.startTransition(() => {
+    const nav = getBrowserNavigationTraversalHints();
+    if (nav !== null) {
+      const canTraverse = direction === "back" ? nav.canGoBack : nav.canGoForward;
+      if (canTraverse) window.__VINEXT_ARM_TRAVERSAL_PENDING__?.();
+    }
+    if (direction === "back") window.history.back();
+    else window.history.forward();
+  });
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/shims/navigation.ts:1394]: Folded into the JSDoc rewrite in `bfc6df17`. The block now explicitly calls out that `router.back()` / `router.forward()` outside a `useTransition` get an implicit transition scope — same pattern as `push` / `replace`, deliberate behaviour change vs. pre-PR `history.back()` synchronous semantics, only way `isPending` can latch across the popstate task boundary.

File: packages/vinext/src/shims/navigation.ts [hunk 4]
```diff
@@ -1239,6 +1239,58 @@ export async function navigateClientSide(
...
+ *
+ * Navigation API hint: `canGoBack` / `canGoForward` are synchronous
+ * availability checks. When `false`, `history.back/forward` is a no-op and
+ * no popstate will fire. Arming the pending in that case would leave it
+ * unsettled (until the next unrelated traversal auto-settles it and
+ * produces a misleading isPending period). We skip arming in that case.
+ *
+ * When the Navigation API is unavailable (very old browsers), we fall
+ * back to the bare traversal without arming. This matches Next.js
+ * degraded behavior; no hang, no safety timer required.
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/shims/navigation.ts]: Adopted in `bfc6df17`. The JSDoc now spells out:

- skipping arm without `canGoBack`/`canGoForward` is intentional, not a TODO;
- arming unconditionally would hang `isPending` on the first no-op traversal in degraded environments;
- explicit "do not 'fix' this branch by re-arming" line so future contributors don't undo it.

File: tests/e2e/helpers.ts [hunk 5]
```diff
@@ -11,14 +11,16 @@ export async function waitForHydration(page: Page): Promise<void> {
...
+ * Checks for the root plus the post-commit router readiness flag.
  *
  * Uses expect().toPass() for better error messages on timeout.
  * 10 second timeout matches Next.js hydration expectations.
  */
 export async function waitForAppRouterHydration(page: Page): Promise<void> {
   await expect(async () => {
-    const ready = await page.evaluate(() => Boolean(window.__VINEXT_RSC_ROOT__));
+    const ready = await page.evaluate(
+      () => Boolean(window.__VINEXT_RSC_ROOT__) && window.__VINEXT_APP_ROUTER_READY__ === true,
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply tests/e2e/helpers.ts]: Symmetrized to `Boolean(...)` on both halves in `64daa1a6`.

File: tests/e2e/app-router/nextjs-compat/router-back-forward-pending.spec.ts [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+    const log = await readObserverLog(page, "__backLog");
+    assertNoIdleFlash(log, "b:pending");
+    expect(page.url()).toBe(`${BASE}/nextjs-compat/router-back-forward-pending`);
+  });
+
+  test("router.forward() keeps isPending true until page B commits", async ({ page }) => {
+    await page.goto(`${BASE}/nextjs-compat/router-back-forward-pending`);
+    await waitForAppRouterHydration(page);
+    await expect(page.locator("#page-a-marker")).toBeVisible({ timeout: 10_000 });
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply tests/e2e/app-router/nextjs-compat/router-back-forward-pending.spec.ts:106]: Comment expanded in `64daa1a6` to name the pollution it avoids: a bare `history.back()` is intentional because `router.back()` would arm a pending and the forward observer log would capture the back traversal's `pending`/`idle` entries instead of isolating the forward traversal under test.

File: packages/vinext/src/server/app-browser-entry.ts [hunk 7]
```diff
@@ -663,14 +673,13 @@ function dispatchBrowserTree(
...
-      return;
-    }
-
+    // Resolve the adopted pending (so a useTransition consumer commits) and
+    // also write the concrete state. The unconditional setter heals rapid
+    // overlapping traversals: when two navigateRsc invocations share one
+    // activePendingBrowserRouterState, the older one's stale-id finally can
+    // settle that pending stale, leaving React's slot bound to a wrong-state
+    // promise resolution. The setter rebinds the slot to the correct state.
+    resolvePendingBrowserRouterState(pendingRouterState, action);
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/server/app-browser-entry.ts:726]: This new Opus 4.7 model has a tendency to litter the code w these verbose comments. I've tried many things locally but it still does this, Ant is prob tokenmaxxing.

Do u think it's possible to switch to 5.5 for bonk? In my experience it catches more correctness bugs and less style nits. It's also quite good at Typescript now compared to 5.4

[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/server/app-browser-entry.ts:726]: @james-elicx

[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/server/app-browser-entry.ts:726]: keep getting issues with the model provider and i dont have time to look into it tonight - if you can get the right model id for me to use with bonk / opencode, i can, but i've tried a couple and they're not working.

[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/server/app-browser-entry.ts:726]: I had a look I think the variant was wrong. Max is for Ant models
Xhigh is highest GPT variant

**Comment by NathanDrake2406** [comment]
Bruh

**Comment by james-elicx** [comment]
should we just go back to opus 4.6?

**Comment by NathanDrake2406** [comment]
Give me 5m. Let me open laptop

**Comment by NathanDrake2406** [comment]
> Give me 5m. Let me open laptop

@james-elicx https://github.com/anomalyco/models.dev/pull/1596

We need to wait for this to get merged to use 5.5
Should we experiment with 5.4 now? It's just one number change. For reviewing it's good imo. In substance is close to 4.7

**Comment by james-elicx** [comment]
Sure thing. thanks for looking into it :)

**Comment by james-elicx** [comment]
oh nice you added a custom config for it

**Comment by james-elicx** [comment]
🙄

**Comment by NathanDrake2406** [comment]
god this is anticlimatic I wanted to see 5.5 reviews so badly :((

**Comment by james-elicx** [comment]
:(

**Comment by NathanDrake2406** [comment]
Damn ig no OAI then :( What do u think? 4.7 but then you tune the instruction heavily to make it less verbose on comments or 4.6?

**Comment by james-elicx** [comment]
yeah worth trying that

**Comment by NathanDrake2406** [comment]
PROGRESS!

**Comment by james-elicx** [comment]
it works! :D

**Comment by james-elicx** [comment]
thanks for digging into it

**Comment by NathanDrake2406** [comment]
the revie quality is kinda mid tho :(

**Comment by NathanDrake2406** [comment]
> thanks for digging into it

actually can u check for other PRs as well? Maybe this is just a mature PR

**Comment by james-elicx** [comment]
okay yeah i ran it on the other pr and its pretty disappointing

**Comment by NathanDrake2406** [comment]
> okay yeah i ran it on the other pr and its pretty disappointing

gimme a sec, claude found another nuance (bug) w opencode

**Comment by james-elicx** [comment]
im just going to switch it back to opus 4.7 for now - we can probably improve the prompt it uses for reviews

**Comment by NathanDrake2406** [comment]
> im just going to switch it back to opus 4.7 for now - we can probably improve the prompt it uses for reviews

yeah, best tradeoff at this point. I think the reasoning is off for 5.4, it usually takes a long time on xhigh.

File: packages/vinext/src/server/app-browser-entry.ts [hunk 8]
```diff
@@ -1087,6 +1100,17 @@ function bootstrapHydration(rscStream: ReadableStream<Uint8Array>): void {
...
+        activePendingBrowserRouterState !== null &&
+        !activePendingBrowserRouterState.settled
+      ) {
+        // Popstate path: the shim already armed a pending inside the user's
+        // `React.startTransition` (router.back / router.forward). Adopt it so
+        // the commit resolves the promise React's state slot is suspended on,
+        // keeping `useTransition().isPending` true across the async popstate
+        // task boundary. User-initiated popstates (keyboard, gesture, address
+        // bar) find no active pending and fall through unchanged.
+        pendingRouterState = activePendingBrowserRouterState;
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/server/app-browser-entry.ts]: 😃

**Comment by NathanDrake2406** [comment]
Reminder for Bonk: I'm not adding verbose comments spelling out everything. 
Also, if the PR is good then please remember to approve it formally, it'll save a lot of time for me and the maintainer. Thanks.

**Comment by NathanDrake2406** [comment]
For Bonk: if LGTM pls formally approve

**Comment by NathanDrake2406** [comment]
Not happy w the architecture so I'm visiting this later
